File: r/universalfft.R
#' UniversalFFT — Boteler (2012) conventions for R
#'
#' Wraps stats::fft() to implement the same two transform pairs as the
#' Python and MATLAB packages.
#'
#' R convention (stats::fft):
#'   fft(x)          → X_raw[k] = Σ x[n] exp(-i2π(k-1)(n-1)/N)  (no scaling)
#'   fft(X, inverse=TRUE) → Σ X[k] exp(+i2π(k-1)(n-1)/N)         (no 1/N)
#'
#' So:
#'   CC forward  = fft(x) * dt
#'   CC inverse  = fft(X, inverse=TRUE) * df   where df = 1/(N*dt)
#'   CD forward  = fft(x) / N
#'   CD inverse  = fft(X, inverse=TRUE) / N * N  = fft(X, inverse=TRUE)
#'
#' NOTE: R's fft(inverse=TRUE) does NOT divide by N (unlike numpy.ifft).
#' This actually simplifies the CD inverse: it is exactly fft(X, inverse=TRUE).


# ---- CC forward (Boteler Eq. 21a) ----------------------------------------

#' @param x  Numeric or complex vector of N time-domain samples.
#' @param dt Sampling interval in seconds.
#' @return   Complex vector of length N in FFT-output order.
ufft_cc_forward <- function(x, dt) {
  stats::fft(x) * dt
}


# ---- CC inverse (Boteler Eq. 21b) ----------------------------------------

#' @param X  Complex frequency-domain vector (FFT-output order), length N.
#' @param dt Sampling interval of the original time series (seconds).
#' @return   Complex vector of N reconstructed time-domain samples.
ufft_cc_inverse <- function(X, dt) {
  N  <- length(X)
  df <- 1.0 / (N * dt)
  # R's fft(inverse=TRUE) returns the raw summation (no 1/N division),
  # so we simply multiply by df.
  stats::fft(X, inverse = TRUE) * df
}


# ---- CD forward (Boteler Eq. 22a) ----------------------------------------

#' @param x  Numeric or complex vector of N time-domain samples.
#' @param dt Unused; kept for API symmetry.
#' @return   Complex vector of length N.
ufft_cd_forward <- function(x, dt = NULL) {
  N <- length(x)
  stats::fft(x) / N
}


# ---- CD inverse (Boteler Eq. 22b) ----------------------------------------

#' Raw summation x[n] = Σ X[k] e^{+i2πkn/N}.
#' R's fft(inverse=TRUE) already returns the raw summation without 1/N,
#' so this is a direct call.
#'
#' @param X  Complex frequency-domain vector (CD-scaled), length N.
#' @param dt Unused; kept for API symmetry.
#' @return   Complex vector of N time-domain samples.
ufft_cd_inverse <- function(X, dt = NULL) {
  stats::fft(X, inverse = TRUE)
}


# ---- Frequency bin array -------------------------------------------------

#' @param N  Number of samples.
#' @param dt Sampling interval in seconds.
#' @return   Real vector of length N containing frequency bins in Hz,
#'           in FFT-output order (same as numpy.fft.fftfreq).
ufft_freqs <- function(N, dt) {
  df <- 1.0 / (N * dt)
  k  <- 0:(N - 1)
  f  <- k * df
  f[k >= N %/% 2] <- (k[k >= N %/% 2] - N) * df
  f
}


# ---- Filter via CC pair ---------------------------------------------------

#' Apply a frequency-domain filter to a time series.
#'
#' @param x  Input time-domain vector (N samples).
#' @param H  Transfer function at FFT frequency bins (length N, complex).
#' @param dt Sampling interval in seconds.
#' @return   Complex filtered time-domain vector.
ufft_filter <- function(x, H, dt) {
  X <- ufft_cc_forward(x, dt)
  Y <- X * H
  ufft_cc_inverse(Y, dt)
}


# ---- Low-pass brick-wall transfer function --------------------------------

#' @param f  Frequency bin vector (from ufft_freqs).
#' @param fc Cut-off frequency in Hz.
#' @return   Real vector: 1.0 where |f| <= fc, 0.0 elsewhere.
ufft_lowpass <- function(f, fc) {
  as.numeric(abs(f) <= fc)
}

File: julia/src/UniversalFFT.jl
"""
    UniversalFFT

Boteler (2012) compliant FFT/IFFT wrappers for Julia.

Julia/FFTW.jl convention (what we wrap)
----------------------------------------
    fft(x)   → X_raw[k] = Σ x[n] e^{-i2π(k-1)(n-1)/N}   (no scaling, 1-indexed)
    ifft(X)  → x_raw[n] = (1/N) Σ X[k] e^{+i2π(k-1)(n-1)/N}

Boteler mapping
---------------
    CC forward  = fft(x)   * dt
    CC inverse  = ifft(X)  * N * df  =  ifft(X) / dt    (since N·df = 1/dt)
    CD forward  = fft(x)   / N
    CD inverse  = ifft(X)  * N                           (undo the 1/N)
"""
module UniversalFFT

using FFTW

export fft_cc, ifft_cc, fft_cd, ifft_cd, fft_filter, freqs,
       low_pass_response, sinc_integral_check, next_power_of_two

# ── CC forward  (Boteler Eq. 21a) ───────────────────────────────────────────
"""
    fft_cc(x, dt) -> Vector{ComplexF64}

Continuous–Continuous forward DFT.

    X[k] = Σ x[n] e^{-i2πkn/N} Δt

# Arguments
- `x`  : time-domain samples (real or complex), length N
- `dt` : sampling interval in seconds
"""
function fft_cc(x::AbstractVector, dt::Real)
    return FFTW.fft(complex.(x)) .* dt
end

# ── CC inverse  (Boteler Eq. 21b) ───────────────────────────────────────────
"""
    ifft_cc(X, dt) -> Vector{ComplexF64}

Continuous–Continuous inverse DFT.

    x[n] = Σ X[k] e^{+i2πkn/N} Δf,    Δf = 1/(N·Δt)

FFTW.ifft applies 1/N, so the full scale factor is N·Δf = 1/Δt.
"""
function ifft_cc(X::AbstractVector, dt::Real)
    N  = length(X)
    df = 1.0 / (N * dt)
    return FFTW.ifft(complex.(X)) .* (N * df)   # == ifft(X) / dt
end

# ── CD forward  (Boteler Eq. 22a) ───────────────────────────────────────────
"""
    fft_cd(x, dt=nothing) -> Vector{ComplexF64}

Continuous–Discrete forward DFT.

    X[k] = (1/N) Σ x[n] e^{-i2πkn/N}

`dt` is unused; accepted for API symmetry.
"""
function fft_cd(x::AbstractVector, dt=nothing)
    N = length(x)
    return FFTW.fft(complex.(x)) ./ N
end

# ── CD inverse  (Boteler Eq. 22b) ───────────────────────────────────────────
"""
    ifft_cd(X, dt=nothing) -> Vector{ComplexF64}

Continuous–Discrete inverse DFT (raw summation).

    x[n] = Σ X[k] e^{+i2πkn/N}

FFTW.ifft divides by N, so we undo that factor.
"""
function ifft_cd(X::AbstractVector, dt=nothing)
    N = length(X)
    return FFTW.ifft(complex.(X)) .* N
end

# ── Frequency bin array ───────────────────────────────────────────────────────
"""
    freqs(N, dt) -> Vector{Float64}

FFT frequency bin array in Hz, in FFT-output order (matching numpy.fft.fftfreq).

    f[k] =  k/(N·dt)       for k = 0 … N/2-1
    f[k] = (k-N)/(N·dt)    for k = N/2 … N-1
"""
function freqs(N::Integer, dt::Real)
    df = 1.0 / (N * dt)
    f  = collect(0:N-1) .* df
    half = N ÷ 2
    f[half+1:end] .= (collect(half:N-1) .- N) .* df
    return f
end

# ── Convenience filter ────────────────────────────────────────────────────────
"""
    fft_filter(x, H, dt) -> Vector{ComplexF64}

Apply transfer function `H` (at FFT frequency bins) to `x` via the CC pair.
Either CC or CD pair gives identical results for a filter pair (Boteler §4.1).
"""
function fft_filter(x::AbstractVector, H::AbstractVector, dt::Real)
    length(x) == length(H) || throw(
        DimensionMismatch("x and H must have the same length")
    )
    return ifft_cc(fft_cc(x, dt) .* complex.(H), dt)
end

# ── Utility helpers ───────────────────────────────────────────────────────────
"""
    low_pass_response(f, fc) -> Vector{Float64}

Brick-wall low-pass transfer function: 1.0 where |f| ≤ fc, 0.0 elsewhere.
"""
low_pass_response(f::AbstractVector, fc::Real) = Float64.(abs.(f) .<= fc)

"""
    sinc_integral_check(h, dt; tol=1e-4) -> Bool

Verify that the impulse response integrates to 1.0 (Boteler §4.2).
"""
function sinc_integral_check(h::AbstractVector, dt::Real; tol::Real=1e-4)
    return abs(sum(real.(h)) * dt - 1.0) < tol
end

"""
    next_power_of_two(n) -> Int

Smallest power of two ≥ n.
"""
next_power_of_two(n::Integer) = n <= 1 ? 1 : 2^(ceil(Int, log2(n)))

end # module
